src/azul/service/repository_controller.py

```diff
@@ -224,3 +243,42 @@ def download_file(self,
             }
         else:
             assert False
+
+    def list_sources(self,
+                     catalog: CatalogName,
+                     request: AzulRequest,
+                     ) -> JSONs:
+        authentication = request.authentication
+        plugin = self.repository_plugin(catalog)
+
+        def list_sources() -> Iterable[SourceRef]:
+            try:
+                return plugin.list_sources(authentication)
+            except PermissionError:
+                raise UnauthorizedError
+
+        if authentication is None:
+            # FIXME Determine public snapshots
+            #       https://github.com/DataBiosphere/azul/issues/2978
+            sources = list_sources()
+        else:
+            cache_key = f'{catalog}:{authentication.identity()}'
+            try:
+                source_jsons = self._source_cache_service.get(cache_key)
+            except CacheMiss:
+                sources = list_sources()
+                source_jsons = [source.to_json() for source in sources]
+                self._source_cache_service.put(cache_key, source_jsons)
+            else:
+                sources = [
+                    # Skip ID verification to avoid O(n) round-trips to TDR
+                    plugin.resolve_source(**source_json, verify_id=False)
```

Comment:
This is assumes that the `resolve_source` argument names match the keys that source.to_json emits. See REVIEW commit.